base/math.jl

```diff
@@ -62,11 +62,7 @@ julia> clamp.([11,8,5],10,6) # an example where lo > hi
  10
 ```
 """
-clamp(x::X, lo::L, hi::H) where {X,L,H} =
-    ifelse(x > hi, convert(promote_type(X,L,H), hi),
-           ifelse(x < lo,
-                  convert(promote_type(X,L,H), lo),
-                  convert(promote_type(X,L,H), x)))
+clamp(x, lo, hi) = max(lo, min(hi, x))
```

Comment:
I appreciate the simplicity of this definition, but I'm concerned that this will significantly slower. Could you benchmark before and after?

Reply:
As Stefan noted, the use `ifelse` here is for performance (it can eliminate branches in tight loops). A benchmark with SIMD vectorization is good to have here.

Reply:
But `max` and `min` use `ifelse` too, so they don't introduce new branches.

Reply:
However the new definition does produce a lot more `llvm` code.

```
julia> @code_llvm debuginfo=:none clamp(x,l,u) # original definition

define double @julia_clamp_17348(double, double, double) {
top:
  %3 = fcmp uge double %2, %0
  %.inv = fcmp olt double %0, %1
  %4 = select i1 %.inv, double %1, double %0
  %5 = select i1 %3, double %4, double %2
  ret double %5
}

julia> myclamp(x, lo, hi) = max(lo, min(hi, x));

julia> @code_llvm debuginfo=:none myclamp(x,l,u)

define double @julia_myclamp_17337(double, double, double) {
top:
  %3 = fcmp olt double %0, %2
  %4 = bitcast double %0 to i64
  %5 = bitcast double %2 to i64
  %6 = icmp sgt i64 %5, -1
  %7 = icmp slt i64 %4, 0
  %8 = and i1 %7, %6
  %9 = or i1 %3, %8
  %10 = fcmp ord double %2, 0.000000e+00
  %11 = select i1 %10, double %0, double %2
  %12 = fcmp ord double %0, 0.000000e+00
  %13 = select i1 %12, double %2, double %0
  %14 = select i1 %9, double %11, double %13
  %15 = fcmp ogt double %14, %1
  %16 = bitcast double %14 to i64
  %17 = bitcast double %1 to i64
  %18 = icmp sgt i64 %16, -1
  %19 = icmp slt i64 %17, 0
  %20 = and i1 %19, %18
  %21 = or i1 %15, %20
  %22 = fcmp ord double %1, 0.000000e+00
  %23 = select i1 %22, double %14, double %1
  %24 = fcmp ord double %14, 0.000000e+00
  %25 = select i1 %24, double %1, double %14
  %26 = select i1 %21, double %23, double %25
  ret double %26
}

```

Reply:
There seems to be a performance hit, though I see vectorization instructions in both cases:

```

julia> f(x)=clamp.(x,0.2,0.9);

julia> g(x)=myclamp.(x,0.2,0.9);

julia> x=rand(10^4);

julia> @btime f($x);
  4.992 μs (2 allocations: 78.20 KiB)

julia> @btime g($x);
  6.939 μs (2 allocations: 78.20 KiB)
```
(Hand written versions of `f,g` without broadcasting, show the same results).

Hope this helps!

Reply:
Something like this does not seem to have much performance penalty:
```
function myclamp(x::X, lo::L, hi::H) where {X,L,H}
       isnan(lo) && return NaN
       isnan(hi) && return NaN
           ifelse(x > hi, convert(promote_type(X,L,H), hi),
                  ifelse(x < lo,
                         convert(promote_type(X,L,H), lo),
                         convert(promote_type(X,L,H), x)))
end

julia> @btime g($x);
  5.011 μs (2 allocations: 78.20 KiB)

julia> @btime f($x);
  4.992 μs (2 allocations: 78.20 KiB)
```
The merits of special casing `NaN` in the above implementation is another issue :)

Reply:
`&&` does introduce a branch. Don't know if it's a bad thing here though.

Reply:
How about:

```julia
myclamp(x::Real, l::Real, u::Real) = myclamp(promote(x,l,u)...)

function myclamp(x::Real, l::Real, u::Real)
	xf, lf, uf = float.(promote(x, l, u))
	ifelse(xf < lf, lf,
		ifelse(xf > uf, uf,
			ifelse(lf ≤ xf ≤ uf, xf,
				oftype(xf, NaN))))
end
```

Reply:
`max(lo, min(hi, x))` does feel like the right answer. I don't understand why it's slower.

Reply:
@raminammour's version seems best. I would write it like this:

```julia
function myclamp(x::X, lo::L, hi::H) where {X,L,H}
   T = promote_type(X,L,H)
   isnan(lo) && return convert(T, NaN)
   isnan(hi) && return convert(T, NaN)
   ifelse(x > hi, convert(T, hi),
          ifelse(x < lo,
                 convert(T, lo),
                 convert(T, x)))
end
```

so we return `NaN` of the correct type. With this I get these timing:

```
julia> @btime f($x);
  3.436 μs (2 allocations: 78.20 KiB)

julia> @btime g($x);
  3.752 μs (2 allocations: 78.20 KiB)
```

Though `clamp` is consistently slightly faster than `myclamp`, I'd say the performance hit is negligible and it seems that a cost is inevitable if add a check for `NaN`.

Reply:
Thanks! I'm tackling it.